File: wear/app/src/main/java/com/gearan/watch/sync/GearanSyncEngine.kt
package com.gearan.watch.sync

import com.gearan.watch.util.GearanLog
import kotlinx.serialization.Serializable

/**
 * Incremental sync engine: after initial sync, only dirty categories travel.
 * Both sides track per-category revision + timestamp; the sender emits only
 * categories marked dirty since the last acknowledged revision.
 * Same category set and rules as ios/GearanSyncEngine (wire-compatible JSON).
 */
object SyncCategories {
    const val DEVICE_STATE = "DEVICE_STATE"
    const val BATTERY = "BATTERY"
    const val HEALTH = "HEALTH"
    const val NOTIFICATIONS = "NOTIFICATIONS"
    const val NETWORK = "NETWORK"
    const val SETTINGS = "SETTINGS"
    const val CAPABILITIES = "CAPABILITIES"

    val ALL = listOf(DEVICE_STATE, BATTERY, HEALTH, NOTIFICATIONS, NETWORK, SETTINGS, CAPABILITIES)
}

@Serializable
data class CategoryState(
    val category: String,
    val revision: Long = 0,
    val timestamp: String = "",
)

@Serializable
data class SyncPayload(
    val deviceId: String,
    val changes: List<CategoryChange>,
)

@Serializable
data class CategoryChange(
    val category: String,
    val revision: Long,
    val timestamp: String,
    /** Opaque per-category JSON (e.g. {"percent":84}). Never health/notification content in logs. */
    val dataJson: String = "{}",
)

class GearanSyncEngine {
    private val revisions = SyncCategories.ALL.associateWith { 0L }.toMutableMap()
    private val dirty = LinkedHashSet<String>()
    var lastSync: String? = null
        private set

    /** Mark a category changed; bumps nothing until collected (coalesces bursts). */
    fun markDirty(category: String, nowIso: String): Boolean {
        if (!revisions.containsKey(category)) return false
        dirty.add(category)
        lastPendingTs[category] = nowIso
        return true
    }

    private val lastPendingTs = mutableMapOf<String, String>()

    fun pendingCount(): Int = dirty.size

    /**
     * Collect dirty categories into one payload (each revision = prev + 1).
     * Returns null when nothing changed — no empty syncs on the radio.
     */
    fun collect(deviceId: String, dataFor: (String) -> String = { "{}" }): SyncPayload? {
        if (dirty.isEmpty()) return null
        val changes = dirty.sorted().map { cat ->
            val rev = (revisions[cat] ?: 0L) + 1
            revisions[cat] = rev
            val ts = lastPendingTs[cat] ?: ""
            CategoryChange(cat, rev, ts, dataFor(cat))
        }
        dirty.clear()
        lastPendingTs.clear()
        GearanLog.sync("incremental sync: ${changes.size} categories")
        return SyncPayload(deviceId, changes)
    }

    /** Apply a received payload; returns categories actually newer than local. */
    fun apply(payload: SyncPayload): List<String> {
        val applied = mutableListOf<String>()
        payload.changes.forEach { c ->
            val local = revisions[c.category]
            if (local != null && c.revision > local) {
                revisions[c.category] = c.revision
                applied.add(c.category)
            }
        }
        if (applied.isNotEmpty()) {
            lastSync = applied.joinToString()
            GearanLog.sync("applied ${applied.size} categories")
        }
        return applied
    }

    fun snapshot(): List<CategoryState> =
        revisions.map { (cat, rev) -> CategoryState(cat, rev) }
}

File: ios/Sources/GearanCore/GearanSyncEngine.swift
import Foundation

/// Incremental sync engine (iOS side). Same categories and rules as
/// wear/GearanSyncEngine: per-category revision + timestamp, only dirty
/// categories travel, stale revisions are rejected. Wire-compatible JSON.
public enum SyncCategory: String, CaseIterable, Codable {
    case DEVICE_STATE, BATTERY, HEALTH, NOTIFICATIONS, NETWORK, SETTINGS, CAPABILITIES
}

public struct CategoryChange: Codable, Equatable {
    public var category: String
    public var revision: Int64
    public var timestamp: String
    public var dataJson: String
    public init(category: String, revision: Int64, timestamp: String, dataJson: String = "{}") {
        self.category = category; self.revision = revision
        self.timestamp = timestamp; self.dataJson = dataJson
    }
}

public struct SyncPayload: Codable, Equatable {
    public var deviceId: String
    public var changes: [CategoryChange]
    public init(deviceId: String, changes: [CategoryChange]) {
        self.deviceId = deviceId; self.changes = changes
    }
}

public final class GearanSyncEngine {
    private var revisions: [String: Int64] = [:]
    private var dirty = Set<String>()
    private var pendingTs: [String: String] = [:]
    public private(set) var lastSync: String?

    public init() {
        SyncCategory.allCases.forEach { revisions[$0.rawValue] = 0 }
    }

    @discardableResult
    public func markDirty(_ category: SyncCategory, nowIso: String) -> Bool {
        dirty.insert(category.rawValue)
        pendingTs[category.rawValue] = nowIso
        return true
    }

    public func pendingCount() -> Int { dirty.count }

    /// Collects dirty categories (each revision = prev + 1). Nil when clean —
    /// no empty syncs on the radio.
    public func collect(deviceId: String, dataFor: (String) -> String = { _ in "{}" }) -> SyncPayload? {
        guard !dirty.isEmpty else { return nil }
        let changes = dirty.sorted().map { cat -> CategoryChange in
            let rev = (revisions[cat] ?? 0) + 1
            revisions[cat] = rev
            return CategoryChange(category: cat, revision: rev,
                                  timestamp: pendingTs[cat] ?? "", dataJson: dataFor(cat))
        }
        dirty.removeAll(); pendingTs.removeAll()
        GearanLog.link("incremental sync: \(changes.count) categories")
        return SyncPayload(deviceId: deviceId, changes: changes)
    }

    /// Applies a received payload; returns categories actually newer than local.
    @discardableResult
    public func apply(_ payload: SyncPayload) -> [String] {
        var applied: [String] = []
        for c in payload.changes {
            if let local = revisions[c.category], c.revision > local {
                revisions[c.category] = c.revision
                applied.append(c.category)
            }
        }
        if !applied.isEmpty {
            lastSync = Date().ISO8601Format()
            GearanLog.link("applied \(applied.count) categories")
        }
        return applied
    }
}
